colexec: add native INet support

yuzefovich opened the pull request:
To be filled.

Addresses: #42043.

Epic: None

Release note: None

yuzefovich committed:
colexec: add native INet support

To be filled.

TODO: figure out what to do in mixed-version clusters where old ones
still use datum-backed representation.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102841)
<!-- Reviewable:end -->

yuzefovich committed:
execversion: extract DistSQL version into a new package

The following commits will introduce dependency on this version into
other low-dependency packages, so this extraction is needed to avoid
import cycles.

Release note: None

yuzefovich committed:
WIP on plumbing ctx

yuzefovich committed:
roachtest: extend version-upgrade

This commit extends the version-upgrade test to verify that scanning
a table with random types succeeds across versions. This can be useful,
for example, when vectorized execution engine changes how it supports
a type (natively vs datum-backed vector).

Release note: None

yuzefovich committed:
WIP on only INet

yuzefovich committed:
WIP on ensuring that we have execversion

yuzefovich committed:
WIP on mixed-version handling

yuzefovich committed:
WIP on plumbing ctx into GetDatumToPhysicalFn

yuzefovich committed:
WIP on cfetcher_wrapper support via IndexFetchSpec versioning

yuzefovich committed:
WIP on ensuring version during physical planning

yuzefovich committed:
stats: revert upgrade of hyperloglog library

This commit partially reverts the bump of the `hyperloglog` library from
b8c532e391b70bb225bfe73b11f8eb14a6dbd87e. The library is used in two
places - for unique counter metrics as well as to compute the distinct
count in the table statistics, and the latter usage needs to work across
versions. With the library version bump it became possible to hit an
index out of range panic when performing the table stats collection in
the following case:
- the cluster is in the mixed-version state where some nodes are running
24.3 and prior versions while other are running binaries with the
library version bump
- the coordinator for the job (i.e. the node that has the sample
aggregator) is running 24.3 and prior versions.

If we want to bump the library version, we'll have to do so with
backwards-compatibility in mind for the table stats collection usage.
For now, we simply revert the bump for that usage.

Release note: None

yuzefovich committed:
WIP on using correct exec version based on cluster version

yuzefovich committed:
execversion: refactor DistSQL version

This commit refactors the DistSQL version mechanism to better serve our
needs. When it was first introduced, it was too coarse since a change in
the logic in one processor would make _all_ physical plans not be
distributable in the mixed-version state, so we have avoided performing
the DistSQL version bumps since 23.1 cockroach version. However, we
still need a mechanism to perform execution logic "upgrades".

The main idea is that we will need to keep the code that can execute every
version in `[MinAccepted, Latest]` range, and we’d have execution logic
conditional on the version that is put into `SetupFlowRequest`. The gateway
for the query will be consulting cluster version to decide which
DistSQL exec version is safe to use. `MinAccepted` can only be bumped (and
the corresponding execution code retired / conditionals removed) once
compatibility with the cluster version that uses it is no longer needed.

For example, right now DistSQL version is 71, and it’s what all 24.3 and
prior binaries use; say in 25.1 we introduce a change in the execution engine,
then we bump the version to 72 but keep MinAccepted as 71. In mixed-version
24.3-25.1 cluster we’ll be using version 71 so that all nodes can participate
in the distributed plan. Once compatibility with 24.3 is no longer needed, we
can bump MinAccepted to 72 and remove the old execution code.

In order to reduce the amount of plumbing necessary for these
migrations, we'll store the DistSQL exec version in the context. This
commit already bumps the 25.1 exec version to 72 (in anticipation of its
usage in the following commit).

Release note: None

yuzefovich committed:
rowexec: upgrade hyperloglog for table stats processors

This commit utilizes the updated DistSQL version mechanism to upgrade
the hyperloglog library used for the distinct count estimation in the
sampler and sample aggregator. In `sketchInfo` struct we now have fields
corresponding to the old and the new versions and we populate only one,
based on the DistSQL version provided on the processor creation via the
context. Only one of the two fields is updated throughout the execution,
so there is no confusion when unmarshalling the sketch datum.

It turns out that we already had a sketch version baked into the sketch
proto, so we could have used that to perform this upgrade. However, we
should rely on the universal DistSQL version mechanism instead, so this
commit also begins the process of removing the sketch version from the
proto by deleting the version checks.

Release note: None

yuzefovich committed:
sql: stop storing statement in the context

We currenntly store the statement in the context as a value in the hopes
of enriching the sentry reports. Anecdotally, most reports don't get the
stmt information (because the current mechanism is missing some code
paths), and even when it's present in most cases the report is still
unactionable. A more lengthy discussion can be found in this slack
[thread](https://cockroachlabs.slack.com/archives/C01RX2G8LT1/p1733954812725919).

This commit removes this altogether in order to free up a spot in the
"fast values" context storage that will be utilized by the following
commit. We also try to include the plan gist into sentry reports, and
I find that a bit more helpful, so I chose to keep plan gists over the
stmts.

Release note: None
